ansible-ipi-install/roles/network-discovery/tasks/main.yml

```diff
@@ -34,24 +34,56 @@
   with_items:
     - "{{ ansible_eligible_interfaces }}"
 
-- name: Get the provisioning interface
-  set_fact:
-    prov_nic: "{{ item }}"
-  when: ansible_{{ item }}.macaddress is defined and ansible_{{ item }}.type == "ether" and ansible_{{ item }}.macaddress == prov_mac
-  with_items:
-    - "{{ ansible_eligible_interfaces }}"
-
 - name: Set lab_pub_nics for master
   set_fact:
-    lab_pub_nics: "{{ [lab_pub_nic] }}"
+    lab_pub_nics: []
 
-- name: Setting public nics for all workers
+- name: Setting public and prov nics 
   block:
-    - set_fact:
+    - name: Set master public nics
+      set_fact:
+        lab_pub_nics:  "{{ lab_pub_nics + [ lab_vars['machine_types'][item.0][item.1]['pub_nic'] ]  }}"
+      with_together: 
+        - "{{ master_vendors }}"
+        - "{{ master_types }}"
+
+    - name: Set worker public nics
+      set_fact:
         lab_pub_nics:  "{{ lab_pub_nics + [ lab_vars['machine_types'][item.0][item.1]['pub_nic'] ]  }}"
       with_together: 
         - "{{ worker_vendors }}"
         - "{{ worker_types }}"
+
+    - name: Set the provisioning interface
+      set_fact:
+        prov_nic: "{{ item }}"
+      with_items:
+        - "{{ ansible_eligible_interfaces }}"
+      when: 
+        - master_types | unique == [provisioner_type]
+        - ansible_{{ item }}.macaddress is defined and ansible_{{ item }}.type == "ether" and ansible_{{ item }}.macaddress == prov_mac
+
+    - block:       
+        - name: Set provisioning interface
+          set_fact:
+            prov_nic: "{{ lab_vars['machine_types'][item.0][item.1]['prov_nic_f04'] if 'r640' in master_fqdns[0].split('.')[0].split('-') and 'f04' in master_fqdns[0].split('.')[0].split('-') else lab_vars['machine_types'][item.0][item.1]['prov_nic'] }}"
```

Comment:
prov_nic is also used to create the bridge on provisioner: https://github.com/redhat-performance/JetSki/blob/master/ansible-ipi-install/roles/node-prep/tasks/40_bridge.yml

So you always need to know the prov_nic on the provisioner even in the case of non-homogeneous deploy of provisioner and masters.

Reply:
Added this task again and updated task to set masters_prov_nic fact.